File: sdk/verifiers/rust/src/receipt.rs
use crate::signing::{
    normalize_evidence_receipt, verify_receipt_with_options, UNPINNED_RECEIPT_BANNER,
};
use crate::types::ReceiptReport;
use crate::util::{
    parse_json_text, reject_duplicate_keys, resolve_signer_key, string_at, u64_at, Result,
    VerifierError,
};
use serde_json::Value;
use std::fs;
use std::path::PathBuf;

pub fn run_receipt(
    pathname: &str,
    signer_key: &str,
    allow_unpinned: bool,
) -> Result<ReceiptReport> {
    let clean = PathBuf::from(pathname);
    let key_hex = resolve_signer_key(signer_key)?;
    // Read the raw text so duplicate-key detection sees every key occurrence,
    // not the last-wins map serde_json would build.
    let text = fs::read_to_string(&clean)
        .map_err(|err| VerifierError::Runtime(format!("read {}: {err}", clean.display())))?;
    let mut report = ReceiptReport {
        path: clean.display().to_string(),
        valid: false,
        unpinned: None,
        action_id: None,
        verdict: None,
        transport: None,
        signer_key: None,
        policy_hash: None,
        chain_seq: None,
        error: None,
    };
    // Reject duplicate object keys before parsing or populating report
    // metadata. Last-wins parsing would otherwise let attacker-controlled
    // duplicate values leak into the displayed rejection report.
    if let Err(err) = reject_duplicate_keys(&text) {
        report.error = Some(err.to_string());
        return Ok(report);
    }
    let receipt: Value = parse_json_text(&text, "malformed JSON")?;
    if string_at(&receipt, &["record_type"]) == Some("evidence_receipt_v2") {
        report.action_id = string_at(&receipt, &["event_id"]).map(str::to_string);
        report.verdict = string_at(&receipt, &["payload", "verdict"]).map(str::to_string);
        report.transport = string_at(&receipt, &["payload", "transport"]).map(str::to_string);
        report.signer_key = Some(key_hex.clone());
        report.policy_hash = string_at(&receipt, &["policy_hash"]).map(str::to_string);
        report.chain_seq = u64_at(&receipt, &["chain_seq"]);
    } else {
        report.action_id = string_at(&receipt, &["action_record", "action_id"]).map(str::to_string);
        report.verdict = string_at(&receipt, &["action_record", "verdict"]).map(str::to_string);
        report.transport = string_at(&receipt, &["action_record", "transport"]).map(str::to_string);
        report.signer_key = string_at(&receipt, &["signer_key"]).map(str::to_string);
        report.policy_hash =
            string_at(&receipt, &["action_record", "policy_hash"]).map(str::to_string);
        report.chain_seq = u64_at(&receipt, &["action_record", "chain_seq"]);
    }
    let is_v2 = string_at(&receipt, &["record_type"]) == Some("evidence_receipt_v2");
    if key_hex.is_empty() && is_v2 {
        match normalize_evidence_receipt(&receipt) {
            Ok(()) => {
                report.valid = allow_unpinned;
                report.unpinned = Some(true);
                report.error = Some(UNPINNED_RECEIPT_BANNER.to_string());
            }
            Err(err) => report.error = Some(err),
        }
        return Ok(report);
    }
    match verify_receipt_with_options(&receipt, &key_hex, allow_unpinned) {
        Ok(()) => {
            report.valid = true;
            if key_hex.is_empty() {
                report.unpinned = Some(true);
                report.error = Some(UNPINNED_RECEIPT_BANNER.to_string());
            }
        }
        Err(err) => report.error = Some(err),
    }
    Ok(report)
}

File: sdk/verifiers/ts/src/receipt.ts
// Copyright 2026 Pipelock contributors
// SPDX-License-Identifier: Apache-2.0

import { readFileSync } from "node:fs";
import * as path from "node:path";
import type { Receipt } from "./types.js";
import { normalizeEvidenceReceipt, unpinnedReceiptBanner, verifyReceipt } from "./signing.js";
import { validateV1Receipt } from "./strict.js";
import { decodeUTF8, parseJSON, rejectDuplicateKeys, resolveSignerKey } from "./util.js";

export interface ReceiptReport {
  path: string;
  valid: boolean;
  unpinned?: boolean;
  action_id?: string;
  verdict?: string;
  transport?: string;
  signer_key?: string;
  policy_hash?: string;
  chain_seq?: number;
  error?: string;
}

export async function runReceipt(
  pathname: string,
  signerKey: string,
  allowUnpinned = false,
): Promise<ReceiptReport> {
  const clean = path.normalize(pathname);
  const keyHex = resolveSignerKey(signerKey);
  const report: ReceiptReport = {
    path: clean,
    valid: false,
  };
  let text: string;
  try {
    text = decodeUTF8(readFileSync(clean), "receipt json");
  } catch (err) {
    report.error = (err as Error).message;
    return report;
  }
  try {
    // Reject duplicate object keys before parsing or populating report
    // metadata. Last-wins parsing would otherwise let attacker-controlled
    // duplicate values leak into the displayed rejection report.
    rejectDuplicateKeys(text);
  } catch (err) {
    report.error = (err as Error).message;
    return report;
  }
  const receipt = parseJSON<Receipt>(text, "receipt json");
  // EV2-FU-1: reject unknown fields on a signed v1 receipt (the v2 evidence
  // receipt has its own schema). The only tolerated unknown surface is the
  // top-level ext bag.
  if (receipt.record_type !== "evidence_receipt_v2") {
    try {
      validateV1Receipt(receipt);
    } catch (err) {
      report.error = (err as Error).message;
      return report;
    }
  }
  if (receipt.record_type === "evidence_receipt_v2") {
    const payload = receipt.payload as { verdict?: string; transport?: string } | undefined;
    report.action_id = receipt.event_id;
    report.verdict = payload?.verdict;
    report.transport = payload?.transport;
    report.signer_key = keyHex;
    report.policy_hash = receipt.policy_hash;
    report.chain_seq = receipt.chain_seq;
  } else {
    report.action_id = receipt.action_record?.action_id;
    report.verdict = receipt.action_record?.verdict;
    report.transport = receipt.action_record?.transport;
    report.signer_key = receipt.signer_key;
    report.policy_hash = receipt.action_record?.policy_hash;
    report.chain_seq = receipt.action_record?.chain_seq;
  }
  try {
    if (keyHex === "" && receipt.record_type === "evidence_receipt_v2") {
      normalizeEvidenceReceipt(receipt);
      report.unpinned = true;
      report.error = unpinnedReceiptBanner;
      report.valid = allowUnpinned;
      return report;
    }
    await verifyReceipt(receipt, keyHex, { allowUnpinned });
    if (keyHex === "") {
      report.unpinned = true;
      report.error = unpinnedReceiptBanner;
    }
    report.valid = true;
  } catch (err) {
    report.error = (err as Error).message;
    if (keyHex === "" && report.error === unpinnedReceiptBanner) {
      report.unpinned = true;
    }
  }
  return report;
}
